optuna/samplers/_nsga2/crossover.py

```diff
@@ -0,0 +1,456 @@
+
+    c1 = 0.5 * ((xs_min + xs_max) - betaq1 * xs_diff)  # Equation (4)
+    c2 = 0.5 * ((xs_min + xs_max) + betaq2 * xs_diff)  # Equation (5)
+
+    # SBX applies crossover with establishment 0.5, and with probability 0.5,
+    # the gene of the parent individual is the gene of the child individual.
+    # The original SBX creates two child individuals,
+    # but optuna's implementation creates only one child individual.
+    # Therefore, when there is no crossover,
+    # the gene is selected with equal probability from the parent individuals x1 and x2.
+
+    child = []
+    for c1_i, c2_i, x1_i, x2_i in zip(c1, c2, xs[0], xs[1]):
+        if rng.rand() < 0.5:
+            if rng.rand() < 0.5:
+                child.append(c1_i)
+            else:
+                child.append(c2_i)
+        else:
+            if rng.rand() < 0.5:
+                child.append(x1_i)
+            else:
+                child.append(x2_i)
+    return np.array(child)
+
+
+def _vsbx(
+    x1: np.ndarray,
+    x2: np.ndarray,
+    rng: np.random.RandomState,
+    eta: float,
+) -> np.ndarray:
+    # https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.422.952&rep=rep1&type=pdf
+    # Section 3.2 Crossover Schemes (vSBX)
+
+    assert x1.shape == x2.shape
+
+    us = rng.uniform(0, 1, size=len(x1))
+    beta_1 = np.power(1 / 2 * us, 1 / (eta + 1))
+    beta_2 = np.power(1 / 2 * (1 - us), 1 / (eta + 1))
+    mask = us > 0.5
+    c1 = 0.5 * ((1 + beta_1) * x1 + (1 - beta_1) * x2)
+    c1[mask] = 0.5 * ((1 - beta_1) * x1 + (1 + beta_1) * x2)[mask]
+    c2 = 0.5 * ((3 - beta_2) * x1 - (1 - beta_2) * x2)
+    c2[mask] = 0.5 * (-(1 - beta_2) * x1 + (3 - beta_2) * x2)[mask]
+
+    # vSBX applies crossover with establishment 0.5, and with probability 0.5,
+    # the gene of the parent individual is the gene of the child individual.
+    # The original SBX creates two child individuals,
+    # but optuna's implementation creates only one child individual.
+    # Therefore, when there is no crossover,
+    # the gene is selected with equal probability from the parent individuals x1 and x2.
+
+    child = []
+    for c1_i, c2_i, x1_i, x2_i in zip(c1, c2, x1, x2):
+        if rng.rand() < 0.5:
+            if rng.rand() < 0.5:
+                child.append(c1_i)
+            else:
+                child.append(c2_i)
+        else:
+            if rng.rand() < 0.5:
+                child.append(x1_i)
+            else:
+                child.append(x2_i)
+    return np.array(child)
+
+
+def _undx(
+    x1: np.ndarray,
+    x2: np.ndarray,
+    x3: np.ndarray,
+    rng: np.random.RandomState,
+    sigma_xi: float,
+    sigma_eta: float,
+) -> np.ndarray:
+    # https://ieeexplore.ieee.org/document/782672
+    # Section 2 Unimodal Normal Distribution Crossover
+
+    assert x1.shape == x2.shape == x3.shape
+
+    def _normalized_x1_to_x2(x1: np.ndarray, x2: np.ndarray) -> np.ndarray:
```

Comment:
I think we should leave comments for these utility functions for code readability.

Reply:
Especially, for "psl".